src/render/src/lib.rs

```diff
@@ -143,3 +148,80 @@ pub mod pso;
 pub mod shade;
 /// Convenience macros
 pub mod macros;
+*/
+
+use std::ops;
+use std::sync::mpsc;
+
+pub struct Gpu<B: Backend> {
+    device: Device<B>,
+    queue: CommandQueue<B, core::General>,
+    garbage: handle::GarbageReceiver<B>,
+}
+
+impl<B: Backend> Gpu<B> {
+    pub fn new<A: core::Adapter<B>>(adapter: &A) -> Self {
+        let queue_descs = adapter.get_queue_families()
+            .iter()
+            .map(|&(ref family, qtype)| (family, qtype, family.num_queues()) )
+            .collect::<Vec<_>>();
+        
+        let core::Gpu { device, mut general_queues, heap_types, memory_heaps, .. } =
+            adapter.open(&queue_descs);
+        
+        let queue = general_queues.remove(0);
+
+        let (garbage_sender, garbage_receiver) = handle::garbage_channel();
+
+        Gpu {
+            device: Device {
+                device,
+                heap_types,
+                memory_heaps,
+                garbage: garbage_sender,
+            },
+            queue,
+            garbage: garbage_receiver,
+        }
+    }
+
+    pub fn cleanup(&mut self) {
+        use core::Device;
+
+        let dev = &mut self.device.device;
+        for garbage in self.garbage.try_iter() {
+            use handle::Garbage::*;
+            match garbage {
+                ShaderLib(sl) => dev.destroy_shader_lib(sl),
```

Comment:
We lack documentation for the `core` api so you probably aren't aware of it but resources need to _externally synchronized_ (in vulkan terms) on destruction. This means that they (or derivates like image views) are not in flight anymore/process by the GPU on destruction. This requires fencing to create a sync point between CPU-GPU. To hide costs and reduce blocking by fences the usual approach is to have basically ringbuffer of 'frames' (e.g size of 3). The approach then would be to re-use existing fences from these frames and cleanup at frame boundaries with a bit of delay: e.g queue up for destruction at frame 0, waiting until beginning of the next frame 0 to destroy. This gives the GPU 2 frames time to catch up and signal frame 0 fence.
The concept extends to command buffers/pools as well. In our `quad` example we wait til the GPU idles after every frame which is slow but currently more a testing ground than best practice.

materials:
- http://developer.download.nvidia.com/gameworks/events/GDC2016/Vulkan_Essentials_GDC16_tlorach.pdf (slide 25, graphical visualization)
- https://www.khronos.org/assets/uploads/developers/library/2016-siggraph/3D-BOF-SIGGRAPH_Jul16.pdf (slide 54, example approach from UE4)

Reply:
Thanks for the explanation. I think holding resources handles until synchronisation has been done would be a good way to ensure that cleanup is not done early.